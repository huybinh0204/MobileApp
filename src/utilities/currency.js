/**
 * @param {number} amount is a amount to transform
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat
 */
export function formatCurrency(amount) {
  if (typeof amount !== 'number') {
    throw new Error('formatCurrency error: Value must be a number');
  }

  return new Intl.NumberFormat('en-US', { currency: 'USD', style: 'currency' }).format(amount);
}

export const transactionTypes = {
  CREDIT: 'CREDIT',
  DEBIT: 'DEBIT',
};
