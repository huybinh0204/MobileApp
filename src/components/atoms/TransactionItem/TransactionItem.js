import PropTypes from 'prop-types';
import React from 'react';
import strings from '_localization';
import { formatCurrency, transactionTypes } from '_utilities/currency';
import { castToLocalTZ } from '_utilities/date';
import {
  Container,
  TitleContainer,
  TransactionAmount,
  TransactionSub,
  TransactionTitle,
} from './TransactionItem.styles';

const transactionState = {
  SETTLED: 'SETTLED',
  UNSETTLED: 'UNSETTLED',
};

const TransactionItem = ({ transaction, isLastItem }) => {
  const { id, details, createdDateTime, type, amount, accountTransactionState } = transaction;

  const amountPrefix = type === transactionTypes.CREDIT ? '+' : '-';
  const formattedAmount = `${amountPrefix}${formatCurrency(amount)}`;
  const processing =
    accountTransactionState === transactionState.UNSETTLED
      ? ` - ${strings.accounts.processing}`
      : '';

  return (
    <Container key={id} isLastItem={isLastItem}>
      <TitleContainer>
        <TransactionTitle>{details}</TransactionTitle>
        <TransactionSub>{`${castToLocalTZ(createdDateTime)}${processing}`}</TransactionSub>
      </TitleContainer>
      <TransactionAmount type={type}>{formattedAmount}</TransactionAmount>
    </Container>
  );
};

TransactionItem.propTypes = {
  transaction: PropTypes.shape({
    id: PropTypes.string,
    details: PropTypes.string,
    createdDateTime: PropTypes.string,
    type: PropTypes.string,
    amount: PropTypes.number,
    accountTransactionState: PropTypes.string,
  }).isRequired,
  isLastItem: PropTypes.bool,
};

TransactionItem.defaultProps = {
  isLastItem: false,
};

export default TransactionItem;
