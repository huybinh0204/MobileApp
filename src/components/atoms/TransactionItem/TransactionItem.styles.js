import styled from '@emotion/native';
import BaseText from '_components/atoms/BaseText/BaseText';
import { transactionTypes } from '_utilities/currency';

export const BottomActionText = styled(BaseText)(({ theme }) => ({
  ...theme.typography.subContentSemiBold,
  color: theme.colors.alpha500,
}));

// Transaction render Item.
export const TransactionTitle = styled(BaseText)(({ theme }) => ({
  ...theme.typography.subContentSemiBold,
  color: theme.colors.beta900,
  marginVertical: theme.spacing.xs,
}));

export const TransactionSub = styled(BaseText)(({ theme }) => ({
  ...theme.typography.tinyContentRegular,
  color: theme.colors.beta400,
}));

export const TransactionAmount = styled(BaseText)(({ theme, type }) => {
  const isCredit = type === transactionTypes.CREDIT;
  const fontStyles = isCredit
    ? theme.typography.subContentSemiBold
    : theme.typography.subContentRegular;
  return {
    ...fontStyles,
    color: isCredit ? theme.colors.alpha500 : theme.colors.beta900,
    marginTop: theme.spacing.xs,
  };
});

export const Container = styled.View(({ theme, isLastItem }) => ({
  flex: 1,
  flexDirection: 'row',
  justifyContent: 'space-between',
  alignItems: 'center',
  marginVertical: theme.spacing.xs,
  borderBottomColor: isLastItem ? theme.colors.transparent : theme.colors.beta50,
  borderBottomWidth: 1,
  paddingBottom: theme.spacing.s,
}));

export const BoderItem = styled.View(({ theme }) => ({
  marginVertical: theme.spacing.s,
  borderBottomColor: theme.colors.beta50,
  borderBottomWidth: 1,
}));

export const TitleContainer = styled.View(({ theme }) => ({
  flex: 1,
  paddingRight: theme.spacing.s,
}));
