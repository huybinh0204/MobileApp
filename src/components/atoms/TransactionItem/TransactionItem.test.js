import { render } from '@testing-library/react-native';
import React from 'react';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import { formatCurrency } from '_utilities/currency';
import TransactionItem from './TransactionItem';

describe('TransactionItem', () => {
  it('should match the snapshot', () => {
    const props = {
      transaction: {
        type: 'CREDIT',
        details: 'test credit',
        amount: 200,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
    };

    const { toJSON } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should match the snapshot being the last item', () => {
    const props = {
      isLastItem: true,
      transaction: {
        type: 'CREDIT',
        details: 'test credit',
        amount: 200,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
    };

    const { toJSON } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should display a credit transaction item', () => {
    const props = {
      isLastItem: false,
      transaction: {
        type: 'CREDIT',
        details: 'test credit',
        amount: 200,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
    };

    const { getByText } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    const amount = getByText(`+${formatCurrency(props.transaction.amount)}`);

    expect(amount).toBeTruthy();
  });

  it('should display a debit transaction item', () => {
    const props = {
      transaction: {
        type: 'DEBIT',
        details: 'test credit',
        amount: 99,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
      isLastItem: false,
    };
    const { getByText } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    const amount = getByText(`-${formatCurrency(props.transaction.amount)}`);

    expect(amount).toBeTruthy();
  });

  it('should have the correct created date', async () => {
    const props = {
      isLastItem: false,
      transaction: {
        type: 'DEBIT',
        details: 'test credit',
        amount: 99,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
    };

    const { getByText } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );
    const date = getByText('October 28');

    expect(date).toBeTruthy();
  });

  it('should have the correct amount format', async () => {
    const props = {
      isLastItem: false,
      transaction: {
        type: 'DEBIT',
        details: 'test credit',
        amount: 99.99,
        createdDateTime: '2020-10-28T15:43:18Z',
      },
    };

    const { getByText } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    const amount = getByText('-$99.99');

    expect(amount).toBeTruthy();
  });

  it('should show a processing transaction label', async () => {
    const props = {
      isLastItem: false,
      transaction: {
        type: 'DEBIT',
        details: 'test credit',
        amount: 99.99,
        createdDateTime: '2020-10-28T15:43:18Z',
        accountTransactionState: 'UNSETTLED',
      },
    };

    const { getByText } = render(
      <ComponentWithProviders component={TransactionItem} componentProps={props} />
    );

    const date = getByText('October 28 - Processing');

    expect(date).toBeTruthy();
  });
});
