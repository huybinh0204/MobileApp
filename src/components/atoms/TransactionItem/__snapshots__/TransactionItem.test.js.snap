// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionItem should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        isLastItem={false}
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "borderBottomColor": "#F2F3F4",
                              "borderBottomWidth": 1,
                              "flex": 1,
                              "flexDirection": "row",
                              "justifyContent": "space-between",
                              "marginVertical": 14.5,
                              "paddingBottom": 29,
                            },
                            undefined,
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "flex": 1,
                                "paddingRight": 29,
                              },
                              undefined,
                            ]
                          }
                        >
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 25.5,
                                  "lineHeight": 35.5,
                                  "marginVertical": 14.5,
                                },
                                undefined,
                              ]
                            }
                          >
                            test credit
                          </Text>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#7F8892",
                                  "fontFamily": "Poppins",
                                  "fontSize": 21.5,
                                  "lineHeight": 30.5,
                                },
                                undefined,
                              ]
                            }
                          >
                            October 28
                          </Text>
                        </View>
                        <Text
                          allowFontScaling={false}
                          style={
                            Array [
                              Object {
                                "color": "#3D3397",
                                "fontFamily": "Poppins-SemiBold",
                                "fontSize": 25.5,
                                "lineHeight": 35.5,
                                "marginTop": 14.5,
                              },
                              undefined,
                            ]
                          }
                          type="CREDIT"
                        >
                          +$200.00
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;

exports[`TransactionItem should match the snapshot being the last item 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        isLastItem={true}
                        style={
                          Array [
                            Object {
                              "alignItems": "center",
                              "borderBottomColor": "transparent",
                              "borderBottomWidth": 1,
                              "flex": 1,
                              "flexDirection": "row",
                              "justifyContent": "space-between",
                              "marginVertical": 14.5,
                              "paddingBottom": 29,
                            },
                            undefined,
                          ]
                        }
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "flex": 1,
                                "paddingRight": 29,
                              },
                              undefined,
                            ]
                          }
                        >
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 25.5,
                                  "lineHeight": 35.5,
                                  "marginVertical": 14.5,
                                },
                                undefined,
                              ]
                            }
                          >
                            test credit
                          </Text>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#7F8892",
                                  "fontFamily": "Poppins",
                                  "fontSize": 21.5,
                                  "lineHeight": 30.5,
                                },
                                undefined,
                              ]
                            }
                          >
                            October 28
                          </Text>
                        </View>
                        <Text
                          allowFontScaling={false}
                          style={
                            Array [
                              Object {
                                "color": "#3D3397",
                                "fontFamily": "Poppins-SemiBold",
                                "fontSize": 25.5,
                                "lineHeight": 35.5,
                                "marginTop": 14.5,
                              },
                              undefined,
                            ]
                          }
                          type="CREDIT"
                        >
                          +$200.00
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
